File: apps/admin-v2/src/infrastructure/api/repositories/api-client-resource.repository.ts
import type {
  ClientResourceRepository,
  CreateClientResourceData,
  ClientBillingProfile,
} from '@/domain/repositories/client-resource.repository';
import type { ClientResource } from '@/domain/entities/client-resource';
import type { PaginatedResult } from '@/shared/types/api';
import api from '../client';
import { mapClientResource } from '../mappers/client-resource.mapper';
import { mapPaginatedResponse } from '../mappers/pagination';

export class ApiClientResourceRepository implements ClientResourceRepository {
  async getAll(
    page?: number,
    search?: string,
    withDebt?: boolean,
  ): Promise<PaginatedResult<ClientResource>> {
    const params: Record<string, unknown> = {};
    if (page) params.page = page;
    if (search) params.search = search;
    // Sólo cuando está activo: mandarlo siempre haría que el backend
    // evaluara el filtro en cada listado.
    if (withDebt) params.with_debt = 1;

    const { data: res } = await api.get('/client-resources', { params: { ...params, all: 1 } });
    return mapPaginatedResponse(res, mapClientResource);
  }

  async getById(id: string): Promise<ClientResource> {
    const { data: res } = await api.get(`/client-resources/${id}`);
    // Response is the resource directly (no data wrapper for single resource)
    return mapClientResource(res.id ? res : res.data);
  }

  async create(data: CreateClientResourceData): Promise<ClientResource> {
    const body: Record<string, unknown> = {};
    if (data.clientId) body.client_id = data.clientId;
    if (data.data) body.data = data.data;
    if (data.plate) body.plate = data.plate;
    if (data.brand) body.brand = data.brand;
    if (data.model) body.model = data.model;
    if (data.color) body.color = data.color;
    if (data.type) body.type = data.type;
    if (data.billingProfile) {
      body.billing_profile = {
        doc_type:   data.billingProfile.docType,
        doc_number: data.billingProfile.docNumber ?? null,
        legal_name: data.billingProfile.legalName ?? null,
        email:      data.billingProfile.email ?? null,
        address:    data.billingProfile.address ?? null,
        phone:      data.billingProfile.phone ?? null,
      };
    }

    const { data: res } = await api.post('/client-resources', body);
    return mapClientResource(res.id ? res : res.data);
  }

  async update(id: string, data: Partial<CreateClientResourceData>): Promise<ClientResource> {
    const body: Record<string, unknown> = {};
    if (data.clientId !== undefined) body.client_id = data.clientId;
    if (data.data !== undefined) body.data = data.data;
    if (data.plate !== undefined) body.plate = data.plate;
    if (data.brand !== undefined) body.brand = data.brand;
    if (data.model !== undefined) body.model = data.model;
    if (data.color !== undefined) body.color = data.color;
    if (data.type !== undefined) body.type = data.type;

    const { data: res } = await api.patch(`/client-resources/${id}`, body);
    return mapClientResource(res.id ? res : res.data);
  }

  async getHistory(id: string): Promise<unknown[]> {
    const { data: res } = await api.get(`/client-resources/${id}/history`);
    return Array.isArray(res) ? res : (res.data ?? []);
  }

  async getBilling(id: string): Promise<ClientBillingProfile> {
    const { data: res } = await api.get(`/client-resources/${id}/billing`);
    return mapBilling(res.data);
  }

  async updateBilling(id: string, data: ClientBillingProfile): Promise<ClientBillingProfile> {
    const { data: res } = await api.put(`/client-resources/${id}/billing`, {
      doc_type:   data.docType,
      doc_number: data.docNumber,
      legal_name: data.legalName,
      email:      data.email,
      address:    data.address,
      phone:      data.phone,
    });
    return mapBilling(res.data);
  }
}

function mapBilling(raw: Record<string, unknown>): ClientBillingProfile {
  return {
    docType:   (raw.doc_type as ClientBillingProfile['docType']) ?? 'final_consumer',
    docNumber: (raw.doc_number as string) ?? '',
    legalName: (raw.legal_name as string) ?? '',
    email:     (raw.email as string) ?? '',
    address:   (raw.address as string) ?? '',
    phone:     (raw.phone as string) ?? '',
  };
}

File: apps/admin_mobile/lib/infrastructure/api/repositories/api_client_resource_repository.dart
import '../../../domain/entities/client_resource.dart';
import '../../../domain/repositories/client_resource_repository.dart';
import '../../../shared/types/paginated_result.dart';
import '../dio_client.dart';
import '../mappers/client_resource_mapper.dart';
import 'paginated_helper.dart';

class ApiClientResourceRepository implements ClientResourceRepository {
  final DioClient _client;

  ApiClientResourceRepository(this._client);

  @override
  Future<PaginatedResult<ClientResource>> getAll(
      {int? page, String? search}) async {
    final params = <String, dynamic>{};
    if (page != null) params['page'] = page;
    if (search != null) params['search'] = search;

    final response =
        await _client.dio.get('/client-resources', queryParameters: params);
    return extractPaginated(response.data, mapClientResource);
  }

  @override
  Future<ClientResource> getById(int id) async {
    final response = await _client.dio.get('/client-resources/$id');
    final data = response.data['data'] ?? response.data;
    return mapClientResource(data as Map<String, dynamic>);
  }

  @override
  Future<ClientResource> create({
    int? clientId,
    Map<String, dynamic>? data,
    String? plate,
    String? brand,
    String? model,
    String? color,
    String? type,
  }) async {
    final body = <String, dynamic>{};
    if (clientId != null) body['client_id'] = clientId;
    if (data != null) body['data'] = data;
    if (plate != null) body['plate'] = plate;
    if (brand != null) body['brand'] = brand;
    if (model != null) body['model'] = model;
    if (color != null) body['color'] = color;
    if (type != null) body['type'] = type;

    final response = await _client.dio.post('/client-resources', data: body);
    final respData = response.data['data'] ?? response.data;
    return mapClientResource(respData as Map<String, dynamic>);
  }

  @override
  Future<ClientResource> update(
    int id, {
    Map<String, dynamic>? data,
    String? plate,
    String? brand,
    String? model,
    String? color,
    String? type,
  }) async {
    final body = <String, dynamic>{};
    if (data != null) body['data'] = data;
    if (plate != null) body['plate'] = plate;
    if (brand != null) body['brand'] = brand;
    if (model != null) body['model'] = model;
    if (color != null) body['color'] = color;
    if (type != null) body['type'] = type;

    final response =
        await _client.dio.patch('/client-resources/$id', data: body);
    final respData = response.data['data'] ?? response.data;
    return mapClientResource(respData as Map<String, dynamic>);
  }

  @override
  Future<List<dynamic>> getHistory(int id) async {
    final response = await _client.dio.get('/client-resources/$id/history');
    final data = response.data['data'] ?? response.data;
    return data as List<dynamic>;
  }
}
